src/Scope/Subst.hs

```diff
@@ -1,92 +1,131 @@
 -- Adapted from http://dev.stephendiehl.com/fun/006_hindley_milner.html
 {- Substitution code -}
 
-module Scope.Subst (SubT(..), Subst, compose,
+module Scope.Subst (Subst(..), FreeVars(..), STerm(..),
                     Substitutable,
                     substM, subst, substWithCtxt, alphaRename,
                     substTags, substDatatype, freeDatatypes,
-                    FreeVars, freeVars) where
+                    freeVars, FreeTmVars, freeVarLs) where
 import qualified Data.Map as Map
 import qualified Data.Set as Set
+import Data.String (IsString)
+import Data.Maybe (fromMaybe)
 import Control.Monad.RWS.Lazy
+import Struct.Exprs
 import Util.Helpers
 import Scope.Ctxt (Ctxt)
-import Scope.Fresh (newVar)
+import qualified Scope.Ctxt as C
+import Scope.Fresh (newVar, newVars)
 import Struct.Lib
 
 ----------------------------------------
 
-data SubT = SubVar Var | SubTm Term | SubTp Type | SubTg Tag
-  deriving (Eq, Ord, Show)
-type Subst = Map Var SubT
+data STerm = Rename TmVar -- `Rename v` means `TmVarL v unchanged`
```

Comment:
Could you clarify this comment a bit? 

What's the difference between `Rename v` and `Replace (TmVarL v)`?

Is it possible you mean "`Rename v` means 'Change `TmVarL`, but not `UsVar`'"?

Although that's what `SubTerm` currently does, I suspect there would be no problem with changing `UsVar` as well. In that case, I suspect that `Rename v` could be eliminated, and `tmVars` could just be a `Map TmVar Term`.